include/pybind11/pytypes.h

```diff
@@ -2270,16 +2267,9 @@ iterator object_api<D>::end() const {
     return iterator::sentinel();
 }
 template <typename D>
-item_accessor object_api<D>::operator[](handle key) const {
-    return {derived(), reinterpret_borrow<object>(key)};
-}
-template <typename D>
-item_accessor object_api<D>::operator[](object &&key) const {
-    return {derived(), std::move(key)};
-}
-template <typename D>
-item_accessor object_api<D>::operator[](const char *key) const {
-    return {derived(), pybind11::str(key)};
+template <typename T>
+item_accessor object_api<D>::operator[](T &&key) const {
```

Comment:
This actually will break quite a few things and cause some perf issues with a needless temporary. I don't quite get the motivation here either. What is wrong with the current overloads?

Reply:
Additionally, this PR should include tests to demonstrate the new behavior.

Reply:
interestingly, https://github.com/cielavenir/pybind11/commit/c59a10d1db42ffffc7cd44ead7b5734722ae62ca works locally on both pypy2 and pypy3.